### ISSUE TITLE
Gradient of torch.where not handling nan correctly

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#18287]
```python
import torch
import math

x = torch.tensor([0.0, math.inf], requires_grad=True)
y = torch.where(x < 1.0, torch.cos(x), torch.zeros_like(x))

print(torch.autograd.grad(y.sum(), x))
```

Outputs: `tensor([-0., nan])`

But since `x[1]` is not `<1.0`, so its value is not used at all, the result should be `tensor([-0., 0.])` instead.

**Comment by SsnL** [comment]
the good old mask problem with the autograd looking only at each step..

**Comment by zasdfgbnm** [comment]
@SsnL Hmmm, maybe I'm picking up the wrong place. But recently after updating PyTorch, something previously work for my model stop working by producing a `nan`. See: https://github.com/aiqm/torchani/pull/192
I will take a deeper look.

**Comment by fmassa** [comment]
This is related to https://github.com/pytorch/pytorch/issues/12986 https://github.com/pytorch/pytorch/issues/15506 and https://github.com/pytorch/pytorch/issues/9688

The fundamental issue is that autograd does not distinguish "no gradient" and "zero gradient".

**Comment by zasdfgbnm** [comment]
Closing this since there are many related issues.
